developer_docs/semantic_stage.md

```diff
@@ -0,0 +1,184 @@
+The attribute `SemanticParser._additional_exprs` exists to hold these expressions.
+This object is a list of lists which is initialised and inserted in `_visit_CodeBlock`.
+A list of lists is necessary in case a `CodeBlock` can be found inside another (e.g. `a = [i for i in range]` where a `CodeBlock` contains the `Assign`, but another exists inside the `For` loop).
+
+In order to avoid problems arising from forgetfulness we try to add additional objects in the most general place possible.
+For example, allocation occurs in the function `SemanticParser._assign_lhs_variable`.
+Variable declarations are created in the printer when needed from the scope variables (this allows each language to place the decorators in the most appropriate location).
+
+## Object Tree
+
+All the objects in the Pyccel AST inherit from [`pyccel.ast.basic.Basic`](../pyccel/ast/basic.py).
+This super-class stores information about how the various objects are related.
+This allows the class to provide functions such as `get_user_nodes` (which returns all objects of a given type which use the node), `get_attribute_nodes` (which returns all objects of a given type which are used by the node), `is_attribute_of` (which indicates if the argument is used by the node), `is_user_of` (which indicates if the argument uses the node), and `substitute` (which allows all occurrences of an object in the node to be replaced by a different object).
+See [`pyccel.ast.basic.Basic`](../pyccel/ast/basic.py) for more information about these functions and other useful utility functions.
+
+The tree is constructed in `Basic.__init__` using the `_attribute_nodes` attribute to recognise the names of attributes which must be added to the tree.
+Nevertheless the object tree should be considered in two situations.
+Firstly, if the object is constructed and AST objects are then added to it (e.g. the member function `pyccel.ast.core.CodeBlock.insert2body` used for the garbage collector).
+In this case the new object does not pass through the constructor of its user.
+It is therefore important to call `set_current_user_node` on the new object to update the tree.
+Secondly, if the object contains all necessary information after the syntactic stage (e.g. `pyccel.ast.core.Continue`) we may be tempted to return the object as is.
+However if this were done there would be multiple user nodes from both the semantic and the syntactic stage.
+For example, if we need to have access to the containing function we could do `expr.get_user_nodes(FunctionDef)`.
+We expect that this only returns semantic objects if `expr` is a result of the semantic stage.
+However if objects such as `pyccel.ast.core.Continue` are returned as is, then we would get access to both the syntactic and the semantic versions of the containing function without any way to distinguish between the two.
+To avoid this it is important to call the `pyccel.ast.basic.Basic.clear_user_nodes` function to remove the syntactic objects from the tree.
```

Comment:
When exactly should this function be called?

Reply:
a41dce79